### ISSUE TITLE
Issues with latest release and also with current master branch

### ISSUE THREAD

**Issue by jonagoldman** [ekmungai/eloquent-ifrs#11]
Hi, I have being playing with this library a bit, as you can see from my pull requests.
All test always succeed, but in practice I was not able to use the package.

I have a few issues:
1. Why it is needed to implement an interface just to use a trait? What is the benefit of this pattern?
2. There are collisions between traits for example `IFRSUser` and `Segregating`, both define the `entity()` method.
3. Using the `master` branch I have another issues, for example the `Vat` model requires an `account_id` property that is never set.

This are just some examples, but there are a few more.
I would like to ask if you can please clarify a little bit what is the status of the master branch, and what I am doing wrong that I cant manage to use latest release also.

Thanks!

**Comment by jonagoldman** [comment]
As a side note, I am not an accountant but I like the style of this package and would like to help if I can.
I have a few questions regarding, for example, how a transaction in this package integrates with a transaction in a shop application, let say for example a payment for a product, or a charge of a few items in a restaurant.
I would like to know if it's possible to talk about such things of if this is out of scope for the project.

Thanks.

**Comment by ekmungai** [comment]
Hi Jona,

First of many thanks for your feedback its greatly appreciated. To answer your questions:

1. The pattern is in line with the idea that an interface describes `what` a class can do, and the methods it requires to be able to acomplish that. The traits on the other hand are the actual code of  `how` that class actually does what it can do. Having such code in a trait allows for reuse.
2. You're right. That trait was introduced by another contributor who helped with database table name collisions.
3. The master branch is rather unstable as I am using the package in another project and I keep adding/removing features as and when doing so seems to make sense. The examples in the readme and the documentation are therefore out of date. I've been working on a release schedule that would have a new release at least once a month. The first one is taking a while because the master is no longer backwards compatible with the latest release and will therefore need to be a major release (v3.0.0).

I'll try my best to release the next stable version towards the end of July.

**Comment by ekmungai** [comment]
> As a side note, I am not an accountant but I like the style of this package and would like to help if I can.
I have a few questions regarding, for example, how a transaction in this package integrates with a transaction in a shop application, let say for example a payment for a product, or a charge of a few items in a restaurant.
I would like to know if it's possible to talk about such things of if this is out of scope for the project.

This is precisely the kind of feedback that I require to make the package more useful so always feel free to ask any question you have.

To intergrate with a shop application, you would first need to setup the Bank, Sales Vat Account and Operating Revenue accounts. You'll also need the Sales Vat object. Once this is done, you could perhaps make a Checkout trait with a checkout method as below:

``` 

use IFRS\Transactions\CashSale;

use IFRS\Models\Account;
use IFRS\Models\Vat;

trait Checkout
{
    /**
     * Create product sale.
     */
    public function checkout($product)
    {	
		$bankAccount = Account::find($bankAccountId);
		$salesVatAccount = Account::find($salesVatAccountId);
		$revenueAccount = Account::find($revenueAccountId);
		
		$outputVat = Vat::find($outputVatId);
		
        new CashSale::create([
		'account_id' => $bankAccount->id,
		'date' => Carbon::now(),
		'narration' => "Sale of ".$product->name,
		])->addLineItem(
			new LineItem([
				'vat_id' => $outputVat->id,
				'account_id' => $revenueAccount->id,
				'vat_account_id' => $salesVatAccount->id,
				'description' => "Sale of " . $product->quantity ." ". $product->units . " of " .$product->name,
				'quantity' => $product->quanitity,
				'amount' => $product->price,
			])
		)->post();
    }
}

```

For the restaurant, you could do:

``` 
trait Checkout
{
    /**
     * Create order payment.
     */
    public function checkout($order)
    {
		$bankAccount = Account::find($bankAccountId);
		$salesVatAccount = Account::find($salesVatAccountId);
		$revenueAccount = Account::find($revenueAccountId);
		
		$outputVat = Vat::find($outputVatId);
		
        $cashSale = new CashSale::create([
		'account_id' => $bankAccount->id,
		'date' => Carbon::now(),
		'narration' => "Payment for order ".$order->orderNumber,
		]);
		
		foreach($order->items as $item){
			$cashSale->addLineItem(
				new LineItem([
					'vat_id' => $outputVat->id,
					'account_id' => $revenueAccount->id,
					'vat_account_id' => $salesVatAccount->id,
					'description' => "Sale of " . $item->quantity ." ". $item->units . " of " .$item->name,
					'quantity' => $item->quanitity,
					'amount' => $item->price,
				])
			);
		}
		$cashSale->post();
    }
}
```
Please let me now if the above makes the use case clearer, of if you have further questions.

Cheers.

**Comment by jonagoldman** [comment]
Hi,

1. I agree, my confusion is the class that implements the interface is not really implementing it. The implementation is actually done in the trait. So if I extend the class with my own class, the class is already using the trait, so its already implementing it.
2. I think point 1 might be the source of the collisions. Most models extend `Eloquent\Model` and then they implement an interface using a trait. I need to take deeper look at it but maybe they need to extend a base model first, and that base model will take enforce the implementations? Maybe you already checked this, it's just a thought.
3. I had an issue also with the latest stable release, but I will double check first with a clean repo clone and get back to you.

Thanks.

_PS: Just see you posted an answer to my second comment, will read it soon!_

**Comment by jonagoldman** [comment]
Hi,
on a clean Laravel installation with the latest stable release [v2.0.1](https://github.com/ekmungai/eloquent-ifrs/tree/v2.0.1) and configured following the readme file:

```php
use IFRS\Traits\IFRSUser;
use IFRS\Interfaces\Recyclable;
use IFRS\Interfaces\Segragatable;

class User extends Authenticatable implements Recyclable, Segregatable
{
    use IFRSUser;
}
```

1. First exception is:
```php
Interface 'App\Segregatable' not found
```
_Fix: this is a typo, so I changed it to Segragatable._

2. Then I get this:
```php
Class App\User contains 3 abstract methods and must therefore be declared abstract or implement the remaining methods (IFRS\Interfaces\Recyclable::bootRecycling, IFRS\Interfaces\Recyclable::recycled, IFRS\Interfaces\Segragatable::bootSegragating)
```
_Fix: Add missing traits that implement all methods:_

```php
// ...
use IFRS\Traits\IFRSUser;
use IFRS\Traits\Recycling;
use IFRS\Traits\Segragating;

class User extends Authenticatable implements Recyclable, Segragatable
{
    use IFRSUser, Recycling, Segragating;
}
```

3. Now I get:

```php
Trait method entity has not been applied, because there are collisions with other trait methods on App\User
```

_Fix: The `entity` method is declared both in `IFRSUser` and in `Segragating`, so I removed `IFRSUser`:_

```php
use Recycling, Segragating;
```

4. Now I get this:

```php
Call to undefined method App\User::restoring()
```

And now I'm not sure whats going on, is this a setup issue on the User model?

**Comment by ekmungai** [comment]
Hmm. Thats pretty wierd the restoring() method is inside the Recycling trait. Could you provide the entire stack trace of the error, perhaps there might be some clues in there

**Comment by jonagoldman** [comment]
Ok so the issue is that the `restoring` event only exists for models with the `SoftDeletes` trait, so this will fail:

https://github.com/ekmungai/eloquent-ifrs/blob/94517327d3c0331facdc68526e93c44c69954120/src/Traits/Recycling.php#L50

I added the the `SoftDeletes` trait to my User model and created the `deleted_at` column but now I get this:

```php
Maximum function nesting level of '256' reached, aborting!
```

This happens because the `Segragating` trait applies `EntityScope` to the user:

https://github.com/ekmungai/eloquent-ifrs/blob/94517327d3c0331facdc68526e93c44c69954120/src/Traits/Segragating.php#L31

And then the `EntityScope` tries to get the `User` again here:

https://github.com/ekmungai/eloquent-ifrs/blob/94517327d3c0331facdc68526e93c44c69954120/src/Scopes/EntityScope.php#L25-L32

now we have an infinite loop!

There is an issue here for sure, I just don't know if its because the example in the readme file is outdated, or because a bug.

Besides that, I just noticed there is a User model in the package that only uses the `IFRSUser` trait:

https://github.com/ekmungai/eloquent-ifrs/blob/94517327d3c0331facdc68526e93c44c69954120/src/User.php#L11-L14

What is the this used for? Why it does not implement the interfaces stated in the readme?

**Comment by ekmungai** [comment]
I think you are getting the infinite loop because your App/User model has the Segragatable trait. The User model in the package uses the `IFRSUser` trait rather than the `Segragating` precisely to avoid this situation, as the `IFRSUser` trait provides the association with the entity model without enforcing the scope. The idea is that Users can be used in the rest of the application without necesarily being tied to the entity model in the package.

**Comment by ekmungai** [comment]
I've edited the readme to remove the segragating trait from the User model. You were right, it was a bug in the example. 
> What is the this used for? Why it does not implement the interfaces stated in the readme?

The User model is there as an example of what the same model in the parent example should look like, as well as for use in testing.

**Comment by jonagoldman** [comment]
Updated the User model and that seems fixed.

Now when I follow the example:

```php
$currency = Currency::create([
    "name" => "Euro",
    "currency_code" => "EUR"
]);

$entity = Entity::create([
    "name" => "Example Company",
    "currency_id" => $currency->id
]);

$outputVat = Vat::create([
    'name' => "Standard Output Vat",
    'code' => "O",
    'rate' => 20,
]);
```

When `Vat::create([...])` is called, I get this:
```php
ErrorException
Trying to get property 'id' of non-object
```

The issue seems to be related to a missing entity id. I noticed that:
1. The `entity_id` column in the users table is `null` for my user.
2. The `Vat` model requires an `entity_id` property, but its not specified on creation.

The problem is related to point 1 or 2?

Here is the stacktrace: https://flareapp.io/share/q5YrRRmX#F58

Thanks.

**Comment by ekmungai** [comment]
Ahh I see the problem. The problem has to do with point 1. After creating the `Entity` object you need to asociate the currently logged in user with it via:
```
Auth::user()->update(["entity_id" => $entity->id ]);
``` 
I've updated the readme example to include this. Thanks again for your feedback
